src/components/NewClusterModal.js

```diff
@@ -149,6 +150,66 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
     ]))
   }
 
+  updateCluster(isStopRequired = false) {
+    const { currentCluster, onSuccess } = this.props
+    const { googleProject, clusterName } = currentCluster
+
+    if (isStopRequired) {
+      notify('success', 'To be updated, your runtime will now stop, and then start.', { timeout: 3000 })
+    }
+
+    return onSuccess(
+      Ajax().Clusters.cluster(googleProject, clusterName).update({
+        machineConfig: this.getMachineConfig()
+      }),
+      isStopRequired ? 5000 : 0)
+  }
+
+  //determines whether the changes are applicable for a call to the leo patch endpoint
+  //see this for a diagram of the conditional this implements https://drive.google.com/file/d/1mtFFecpQTkGYWSgPlaHksYaIudWHa0dY/view
+  //this function returns true for cases 2 & 3 in this diagram
+  canUpdate() {
+    const { currentCluster } = this.props
+    const { jupyterUserScriptUri, originalJupyterUserScriptUri, originalImageUrl, selectedLeoImage, customEnvImage } = this.state
+    const currentClusterConfig = currentCluster.machineConfig
+    const userSelectedConfig = this.getMachineConfig()
+
+    const hasStartupScriptChanged = jupyterUserScriptUri !== originalJupyterUserScriptUri
+
+    const hasImageChanged = !_.includes(originalImageUrl, [selectedLeoImage, customEnvImage])
+
+    const workersCantUpdate = currentClusterConfig.numberOfWorkers !== userSelectedConfig.numberOfWorkers &&
+      (currentClusterConfig.numberOfWorkers < 2 || userSelectedConfig.numberOfWorkers < 2)
+
+    const hasUnUpdateableResourceChanged =
+      currentClusterConfig.workerDiskSize !== userSelectedConfig.workerDiskSize ||
+      currentClusterConfig.workerMachineType !== userSelectedConfig.workerMachineType ||
+      currentClusterConfig.numberOfWorkerLocalSSDs !== userSelectedConfig.numberOfWorkerLocalSSDs
+
+    const hasWorkers = currentClusterConfig.numberOfWorkers >= 2 || currentClusterConfig.numberOfPreemptibleWorkers >= 2
+
+    const hasWorkersResourceChanged = hasWorkers && hasUnUpdateableResourceChanged
+
+    const hasDiskSizeDecreased = currentClusterConfig.masterDiskSize > userSelectedConfig.masterDiskSize
+
+    const cantUpdate = workersCantUpdate || hasWorkersResourceChanged || hasDiskSizeDecreased || hasImageChanged || hasStartupScriptChanged || !currentCluster
+    return !cantUpdate
```

Comment:
Nit: It's a bit awkward to calculate a negative and then negate it. Also, this feels like it would more obviously implement the diagram if written as if/else statements. If that's not the case and you want to keep it a set of boolean expressions, it might be easier to digest if the intermediate tests are extracted into separate functions. `hasStartupScriptChanged`, `hasImageChanged`, `workersCantUpdate`, `hasUnUpdateableResourceChanged`, `hasWorkers`, and `hasDiskSizeDecreased` are completely independent of each other. Still, I have a feeling that some of those are only needed some of the time which would be easier to recognize using if/else.

Reply:
I feel like this is the simplest way to do it. It's easier to reason about since each situation is completely isolated in its own boolean, as opposed to needing to reason about a nested structure similar to that in `componentDidMount`. 

I split the state-dependent lines into their own function with the piece of state they need, but I kept the other pieces, all of which rely on the below code in some way. I didn't like how those pieces looked in their own function because of all the duplication in variable initialization steps.

I agree that its atypical to calculate something, and then return its negation, but I found it was the easiest way to make each piece easy to reason about as it resulted in the least amount of negations overall. 

```    
const { currentCluster } = this.props
const currentClusterConfig = currentCluster.machineConfig
const userSelectedConfig = this.getMachineConfig()
```